Units issue with FK4 transforms with and without e terms

mhvk opened the pull request:
Mostly to @eteq: in the process of looking at trying to make coordinates fully shaped, I encountered the `fk4_to_fk4_no_e` transformation (and its inverse) and realised there surely is a units problem in https://github.com/astropy/astropy/blob/master/astropy/coordinates/builtin_frames/fk4.py#L173
```
r = r - eterms_a.reshape(3, 1) + np.dot(eterms_a, r) * r
```
Here, as written, `r` has units of length since it comes from `fk4coord.cartesian.xyz`, but `eterms_a` has units of degrees, since it just multiplies the constant of abberation `k` (which is given as a number that is in degrees, though it is the ratio of the speed of the earth to the speed of light, so is really a unit-less number), with eccentricity `e` (why converted to radians?) and some cosines and sines. I *think* what really should have happened here is to have `r` divided by `d_orig` so that it contains dimensionless direction cosines, but is this true?

My suggestion would be to write `fk4_e_terms` using units, and also use units throughout, so we're sure that what is being done is correct.

(Separately, the normalisation distances do not take into account the possibility of multiple coordinates, but that I can solve trivially if I understand how the above equation is supposed to work.)

mhvk commented:
@eteq, @astrofrog - looking with fresh eyes after a long time at this FK4 e-terms issue, I see that previous results were mostly correct because two errors compensated each other: while the unit of the constant of aberration was inconsistently taken to be in radians, the eccentricity of the earth's orbit was inconsistently taken to be in degrees. Since only the product of the two enters, the final answer was correct. Anyway, I now made the simplest possible correction (still without explicitly using units), and also ensured that the normalisation of distances was done correctly. Please have a look.

p.s. Since the normalisation was a bug that could have affected coordinates with non-unit distances, I'll add a note to the changelog.

mhvk commented:
@StuartLittlefair - just realised that since you fixed some aberration issues earlier, you might also be a logical one to have a look at this PR.

StuartLittlefair commented on astropy/coordinates/builtin_frames/fk4.py:
```diff
@@ -144,11 +144,11 @@ def fk4_e_terms(equinox):
     """
     # Constant of aberration at J2000
-    k = 0.0056932
+    k = 0.0056932  # in degrees (v_earth/c ~ 1e-4 rad ~ 0.0057 deg)
```
For a source, we could cite Explanatory Supplement to the Astronomical Almanac (Seidelmann, 2005) which gives this exact value on p171.

StuartLittlefair commented:
@mhvk this looks right to me.

As a double check I tested the implementation here against the transformations on pp170 of Explanatory Supplement to the Astronomical Almanac (Seidelmann, 2005) which is an independent test of sorts since his formulae work directly on the ```ra``` and ```dec``` properties whilst the astropy routine transforms the ```CartesianRepresentation``` . The two methods agree so I think this is 👍 

I include my test below as a separate comment in case you want to add as a regression test...

StuartLittlefair commented:
Here's my test:

```python
from astropy import coordinates as coord, units as u

fk4 = coord.FK4(10.00039633*u.deg, 40.99492542*u.deg) # M31
equinox = fk4.equinox

Dc = -0.065838*u.arcsec
Dd = 0.35299*u.arcsec

fk4noe_dec = fk4.dec - (Dd*np.cos(fk4.ra) - Dc*np.sin(fk4.ra))*np.sin(fk4.dec) -Dc*np.tan(o)*np.cos(fk4.dec)
fk4noe_ra = fk4.ra - (Dc*np.cos(fk4.ra) + Dd*np.sin(fk4.ra))/(np.cos(fk4.dec)) 

fk4noe = fk4.transform_to(coord.FK4NoETerms)
assert_allclose(fk4noe.ra, fk4noe_ra, rtol=5e-7)
assert_allclose(fk4noe.dec, fk4noe_dec)
```

astrofrog commented:
We should make sure we update the [coordinates-benchmark](https://github.com/astropy/coordinates-benchmark) after this.

mhvk commented:
@StuartLittlefair - thanks!  I added the test. @eteq, @astrofrog, or @adrn - could you please check that the normalisation in the distances is now done correctly? The original equation was clearly wrong, but I'm not sure that my "correction" is OK, since it effectively assumes large distance.

mhvk commented:
I rebased to get rid of the unrelated failures from a new scipy version. I also looked up the explanatory supplement to convince myself that any conversion from FK4 involving e-terms is assumed to be on the unit sphere, which means that the current code should be correct. @eteq, @astrofrog, or @adrn - would you please do a final check?

bsipocz commented:
Could someone change the milestone? 1.0.10 is already released. (Or maybe the friendly astropybot does check it, too?)

mhvk commented:
@bsipocz - thanks for noticing. It is actually a bit odd that 1.0.10 got released without this being noted! Anyway, I moved the entry in the changelog and rebased for good measure.

@eteq, @astrofrog, or @adrn - I think this really should be OK, but could one of you do a final check?

adrn commented on astropy/coordinates/tests/test_regression.py:
```diff
@@ -225,3 +226,30 @@ def test_regression_4996():
+
+def test_regression_4293():
+    """Really just an extra test on FK4 no e, after finding that the units
+    were not always taken correctly.  This test is against explicityly doing
```
explicityly --> explicitly

mhvk replied:
Done

adrn commented on astropy/coordinates/tests/test_regression.py:
```diff
@@ -225,3 +226,30 @@ def test_regression_4996():
+                          Dd*np.sin(fk4.ra)) / (np.cos(fk4.dec))
+
+    fk4noe = fk4.transform_to(FK4NoETerms)
+    assert_quantity_allclose(fk4noe.ra, fk4noe_ra, rtol=5e-7)
```
Is this `rtol` value a "Magic Number," or is it set by some heuristic? Maybe just make an inline comment note about where that came from

mhvk replied:
I had just copied it from https://github.com/astropy/astropy/pull/4293#issuecomment-234973086, but looking closer at it, I see @StuartLittlefair had actually made a typo in `Dd`; with that corrected, the routines match to ~10 μas.

adrn commented:
Also: did you mean to update the `astropy_helpers` commit?

@mhvk Code looks fine to me -- a few minor notes above. Thanks!

mhvk commented:
@adrn - thanks for the review, and for noticing the "abritrarily set" precision of the test. This gives a bit more confidence in the routine itself too.

mhvk commented:
Thanks all for the reviews! Will now merge.
